### ISSUE TITLE
`crucible-mir`: Make translation errors non-fatal

### ISSUE THREAD

**Issue by RyanGlScott** [GaloisInc/crucible#1131]
Consider the following program:

```rs
pub fn f() -> u32 {
    42
}

pub fn g() {
    panic!("Don't run me")
}
```

It is conceivable that we would want to verify `f` in SAW but not `g`. Despite this, if you load this program's MIR JSON into SAW using `mir_load_module`, it will crash before you even have a chance to verify any functions:

```
[12:54:23.144] Loading file "/home/ryanscott/Documents/Hacking/Haskell/saw-script/test.saw"
[12:54:23.180] standalone use of `dyn` is not supported: TyDynamic core/73237d41::any[0]::Any[0]::_trait646b6cc91b1abac7[0]
CallStack (from HasCallStack):
  error, called at src/Mir/TransTy.hs:228:25 in crucible-mir-0.1-inplace:Mir.TransTy
  tyToRepr, called at src/Mir/Trans.hs:868:12 in crucible-mir-0.1-inplace:Mir.Trans
  evalCast', called at src/Mir/Trans.hs:936:5 in crucible-mir-0.1-inplace:Mir.Trans
  evalCast, called at src/Mir/Trans.hs:993:30 in crucible-mir-0.1-inplace:Mir.Trans
  evalRval, called at src/Mir/Trans.hs:1219:11 in crucible-mir-0.1-inplace:Mir.Trans
  transStatement, called at src/Mir/Trans.hs:1685:68 in crucible-mir-0.1-inplace:Mir.Trans
  translateBlockBody, called at src/Mir/Trans.hs:1694:28 in crucible-mir-0.1-inplace:Mir.Trans
  registerBlock, called at src/Mir/Trans.hs:1734:10 in crucible-mir-0.1-inplace:Mir.Trans
  genFn, called at src/Mir/Trans.hs:1780:24 in crucible-mir-0.1-inplace:Mir.Trans
  transDefine, called at src/Mir/Trans.hs:2237:30 in crucible-mir-0.1-inplace:Mir.Trans
  transCollection, called at src/Mir/ParseTranslate.hs:76:26 in crucible-mir-0.1-inplace:Mir.ParseTranslate
  translateMIR, called at src/SAWScript/Crucible/MIR/Builtins.hs:358:11 in saw-script-1.0.0.99-inplace:SAWScript.Crucible.MIR.Builtins
```

Ultimately, the issue lies in the fact that `crucible-mir` crashes when translating `g` into a Crucible CFG, **even if that CFG is never simulated**. This is a poor experience for SAW users, as it means that functions unrelated to the verification effort can prevent the code from being loaded into SAW, and the only workaround at present is to comment out large chunks of code.

When `crucible-mir` encounters a situation like this, it should not crash, but instead emit a CFG that errors when simulated. This is the behavior of `crucible-mir`'s [`mirFail`](https://github.com/GaloisInc/crucible/blob/0d26e4a656176d47dcc9fba9503de3c0d8a0219f/crucible-mir/src/Mir/Generator.hs#L514-L530) function, and the fact that `crucible-mir` crashes on this program is the result of using `error` instead of `mirFail` here:

https://github.com/GaloisInc/crucible/blob/0d26e4a656176d47dcc9fba9503de3c0d8a0219f/crucible-mir/src/Mir/TransTy.hs#L225-L228

We should replace `error` with `mirFail` here, which would likely fix the issue at hand. We should also audit other uses of `error` in `MirGenerator`-adjacent code to make sure that similar issues are prevented.

A secondary goal is to make `crucible-mir` be able to successfully translate code involving `panic!`. Doing so would require figuring out which function(s) in `panic!`'s chain of macro-expanded function calls needs to be overridden. For the time being, a workaround is to use `unimplemented!` instead of `panic!`.

**Comment by spernsteiner** [comment]
In general I agree with changing `error` to `mirFail` wherever we can.  But I'm surprised to see this particular `error` being hit.  "Standalone use of `dyn`" should mean that it saw a local of type `dyn Foo`, which is normally illegal - `dyn Foo` unsized, so it can only be manipulated through a pointer like `&dyn Foo`.  We should investigate whether this error is coming from code in the standard library that's using the unstable "unsized locals" feature, and if so, whether we can patch it out.

**Comment by RyanGlScott** [comment]
Sadly, my idea of replacing `error` with `mirFail` doesn't quite work, since `mirFail` requires the `MirGenerator` monad, but the call to `error` above is is in `tyToRepr`, which is pure. Moreover, the fact that it is pure appears important, as it is used in quite a few contexts where making it return something in `MirGenerator` would be impractical.

**Comment by sauclovian-g** [comment]
User-facing errors really need either a monad or to be returned out explicitly, which sounds like it's going to be a headache. 😿

**Comment by sauclovian-g** [comment]
(or I guess the other option is to check for and reject unsupported stuff further upstream and then convert this error to a panic)

**Comment by RyanGlScott** [comment]
I am increasingly on the side of @spernsteiner's original claim that this translation error shouldn't be happening in the first place. As further evidence of this claim consider what happens if you run a slightly modified version of the same program (the only difference being the addition of `#[crux::test]` attributes) through `crux-mir`:

```rs
#[crux::test]
pub fn f() -> u32 {
    42
}

#[crux::test]
pub fn g() {
    panic!("Don't run me")
}
```

This does not panic, but instead produces reasonable results:

```
$ cabal -v0 run exe:crux-mir -- test.rs 
test test/d0c80282::f[0]: returned 42, ok
test test/d0c80282::g[0]: [Crux] Attempting to prove verification conditions.
FAILED

failures:

---- test/d0c80282::g[0] counterexamples ----
[Crux] Found counterexample for verification goal
[Crux]   internal: error: in test/d0c80282::g[0]
[Crux]   panicking::panic_fmt, called from test/d0c80282::g[0]

[Crux-MIR] ---- FINAL RESULTS ----
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 0
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```

Why, then, does loading the same program with SAW panic? I'm not entirely sure yet, but my suspicion is that `saw-rustc` is producing a MIR JSON file with much more code than it ought to be producing. As evidence for this claim, compare the size of `test.all.mir` (a byproduct of running `mir-json` via `crux-mir`) versus the size of `test.linked-mir.json` (the output of `saw-rustc`):

```
-rw-rw-r--  1 ryanscott ryanscott  69K Feb 26 08:22 test.all.mir
-rw-rw-r--  1 ryanscott ryanscott  53M Feb 26 08:24 test.linked-mir.json
```

The former is a reasonable 69K, but the latter is a whopping 53M? This doesn't make sense, given that in both MIR JSON files, the only roots are the `f` and `g` functions:

```
$ jq '.roots' test.all.mir
[
  "test/d0c80282::f",
  "test/d0c80282::g"
]

$ jq '.roots' test.linked-mir.json 
[
  "test/d84886c2::f",
  "test/d84886c2::g"
]
```

I suspect that this may ultimately be a `mir-json` bug in disguise. I'll investigate further.

**Comment by spernsteiner** [comment]
I agree this is likely a `mir-json` bug.  Calling `panic!` pulls in a lot of code, but I don't think it should be 50 MB worth.  I also don't think `cargo crux-test` and `cargo saw-build` should produce different results when all functions are marked `#[crux::test]` - the main difference between the two is that `cargo saw-build` treats all non-generic functions as if they had `#[crux::test]`.  I wonder if there's some issue in the way that `cargo saw-build` is invoking the mir-json linker that's causing it to keep more symbols than are actually needed.

It might be worth setting up the example code as a cargo project and testing `cargo crux-test`/`cargo saw-build` in addition to `crux-rustc`/`saw-rustc`, in case the build flags passed by Cargo affect the result.

**Comment by RyanGlScott** [comment]
If I compile the program above as a `cargo` project and build then with `cargo crux-test`/`cargo saw-build`, then they are both reasonably sized (and both load into SAW without issues). I suspect that `saw-rustc` in particular is the culprit.

**Comment by RyanGlScott** [comment]
Also, if I run `crux-rustc --test test.rs`, then I get an unreasonably large MIR JSON file, just as I would with `saw-rustc`. I suspect the issue is not Crux versus SAW so much as it is `{crux,saw}-rustc` versus `cargo {crux-test,saw-build}`.
